mlrun/artifacts/base.py

```diff
@@ -26,7 +27,739 @@
+        )
+
+    @property
+    def inline(self):
+        """inline data (body)"""
+
+        if self._is_inline:
+            return self.get_body()
+        return None
+
+    @inline.setter
+    def inline(self, body):
+        self._body = body
+
+    def get_body(self):
+        """get the artifact body when inline"""
+        return self._body
+
+
+class ArtifactStatus(ModelObj):
+    _dict_fields = ["state"]
+
+    def __init__(self):
+        self.state = "created"
+
+    def base_dict(self):
+        return super().to_dict()
+
+
 class Artifact(ModelObj):
+    kind = "artifact"
+    _dict_fields = ["kind", "metadata", "spec", "status"]
+
+    _store_prefix = StorePrefix.Artifact
+
+    def __init__(
+        self,
+        key=None,
+        body=None,
+        viewer=None,
+        is_inline=False,
+        format=None,
+        size=None,
+        target_path=None,
+        # All params up until here are legacy params for compatibility with legacy artifacts.
+        project=None,
+        metadata: ArtifactMetadata = None,
+        spec: ArtifactSpec = None,
+    ):
+        self._metadata = None
+        self.metadata = metadata
+        self._spec = None
+        self.spec = spec
+
+        self.metadata.key = key or self.metadata.key
+        self.metadata.project = (
+            project or mlrun.mlconf.default_project or self.metadata.project
+        )
+        self.spec.size = size or self.spec.size
+        self.spec.target_path = target_path or self.spec.target_path
+        self.spec.format = format or self.spec.format
+        self.spec.viewer = viewer or self.spec.viewer
+
+        if body:
+            self.spec.inline = body
+        self.spec._is_inline = is_inline or self.spec._is_inline
+
+        self.status = ArtifactStatus()
+
+    @property
+    def metadata(self) -> ArtifactMetadata:
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, metadata):
+        self._metadata = self._verify_dict(metadata, "metadata", ArtifactMetadata)
+
+    @property
+    def spec(self) -> ArtifactSpec:
+        return self._spec
+
+    @spec.setter
+    def spec(self, spec):
+        self._spec = self._verify_dict(spec, "spec", ArtifactSpec)
+
+    @property
+    def status(self) -> ArtifactStatus:
+        return self._status
+
+    @status.setter
+    def status(self, status):
+        self._status = self._verify_dict(status, "status", ArtifactStatus)
+
+    def before_log(self):
+        pass
+
+    @property
+    def is_dir(self):
+        """this is a directory"""
+        return False
+
+    @property
+    def inline(self):
+        """This is a property of the spec, look there for documentation
+        leaving here for backwards compatibility with users code that used ArtifactLegacy"""
+        warnings.warn(
+            "This is a property of the spec, use artifact.spec.inline instead"
+            "This will be deprecated in 1.0.0, and will be removed in 1.2.0",
```

Comment:
Since this will only get inside in 1.0.0 I guess the schedule should be changed to 1.2.0 and 1.4.0
In all others as well

Reply:
As we've discussed - since we'll probably not release this before 1.1.0, I've changed this to 1.3.0 and 1.5.0.